basic/x16additions.s

```diff
@@ -954,6 +954,120 @@ cedit:
 	bannex_call bannex_x16edit
 	rts
 
+;******************************************************************
+;
+; LIST [ xxxx - xxxx ]
+; List BASIC program currently in memory and pause after first page
+; Spacebar will pause/unpause
+; In paused mode below will work:
+; PageDown will show one page at a time
+; Arrow down will show one line at a time
+;
+;******************************************************************
+nlines	= $0387			; These variables are in KERNAL space
+llen	= $0386			; Could not figure out how to import
+curs_y	= $0383
+listp:
+	php			; Save cpu flags as they are used after this function
+	pha			; BASIC uses the a,x&y registers, they will be
+	phy			; restored before returning from this function
+	phx
+	stz	ram_bank	; Set RAM bank 0 for variables
+
+	lda	$200		; Use keyboard buffer to see if first run
+	beq	@notfirst
+	stz	$200
+	stz	lp_dopause	; Initialize variables
+	stz	lp_screenpause
+@notfirst:
+	jsr	$FFE4		; GETIN
+	cmp	#$20		; Spacebar
+	bne	@cont
+	inc	lp_dopause
+	jmp	@end
+@cont:	lda	lp_screenpause
+	beq	@islinepause
+	; Handle screen pause
+	jsr	@ateos		; Check if we are at end of screen
+	bcc	@end
+	inc	lp_dopause	; Pause the listing
+	stz	lp_screenpause
+	bra	@end
+@islinepause:
+	lda	lp_dopause	; Check if we need to pause the listing
+	beq	@end
+@pauseloop:
+	jsr	$FFE4		; GETIN
+	cmp	#$03		; Is STOP (CTRL+C)?
+	bne	@space
+	jsr	$FEC3		; Push STOP back in keyboard buffer
```

Comment:
Please `.include "kernal.inc"` and use the symbolic names for kernal calls.  If the call is missing from kernal.inc, there is a clash somewhere.  In that case, please define the call at the top of the source file that contains this function.